torch/distributed/rpc/backend_registry.py

```diff
@@ -254,6 +254,7 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
 
     try:
         _tensorpipe_check_device_maps(agent, rpc_backend_options.device_maps)
+        agent.join()
```

Comment:
This makes me a bit sad, as I thought in principle we wanted to eventually get rid of the `sync` and `join` methods of the agent, and adding more usages will just make it harder. What is the reason for doing this actually? By this time there shouldn't be any RPCs in flight so joining seems to be unnecessary. The only "side effect" I see is that it adds a barrier between all agents. If that's why we use it, can we just make an explicit barrier instead of using join?

Reply:
> This makes me a bit sad, as I thought in principle we wanted to eventually get rid of the sync and join methods of the agent, and adding more usages will just make it harder. 

Agree, I am sad too. 

> What is the reason for doing this actually? By this time there shouldn't be any RPCs in flight so joining seems to be unnecessary. 

There can be RPCs at this point. Assume there are two processes X and Y. Y could finish `init_rpc` earlier than X and then started sending RPCs to X. At this point, without this join, it is possible that X has not configured the reverse device map, and hence hit failures.

> The only "side effect" I see is that it adds a barrier between all agents. If that's why we use it, can we just make an explicit barrier instead of using join?

Which `barrier` are you referring to? The `dist.barrier()`? Will that add another dependency on `c10d`? I cannot tell which one is worse. :)

Reply:
I see, thanks for clarifying that there could be RPCs already in flight, I hadn't realized.

So the problem, at a high level, seems to me that we run in troubles if one rank exits init_rpc early and starts doing stuff while other ranks are potentially still inside init_rpc, right? If that's the case, can't we fix it by doing a barrier at the very end of init_rpc?

On one hand, that would nicely match what @pritamdamania is doing for DDP in #45181 (adding a barrier at the end of init_process_group and new_group). Also, since a barrier is a "weaker" operation than joining the agent (because the latter performs multiple barriers), I'd consider a barrier to be more readable as it makes it clear what the core issue is and what the essential solution is, without going "overkill".

And, if we're doing a barrier, we could use _all_gather and thus avoid using both the (deprecated?) agent.join() and without adding a dependency on c10d.

Reply:
Yep, I would expect `_all_gather` to work too. I actually have one PR for that #44990. Let me land that one too.

Reply:
Using `_all_gather` makes `test_local_shutdown` flaky. Will come back to debug this after branch cut.

Reply:
Sounds good, thanks!